src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -1050,16 +1050,37 @@ public static bool TryDecode(ImmutableArray<byte> bytes)
 
         /// <summary>
         /// Given an input string changes it to be acceptable as a part of a type name.
-        /// For now we will simply replace '.' with '_'as the most common case.
         /// </summary>
         internal static string MangleForTypeNameIfNeeded(string moduleName)
-        {
-            // TODO: it may make sense to strenthen this algorithm 
-            //       to result in 1-1 mapping to reduce chances of
-            //       producing matching results for distinct original strings
-            var result = moduleName.Replace('.', '_');
+        {            
+            var s = new StringBuilder();
```

Comment:
@tmat If we use the pooled string building, I don't think there is value in making the rest of the code conditional.